# Urban Import AI
## Fast Prompting en Acción: Desentrañando la Magia

### Proof of Concept (POC)

**Proyecto Final**

Urban Import AI es una prueba de concepto que utiliza técnicas de
Fast Prompting para mejorar las respuestas de un asistente virtual
orientado a consultas sobre productos de importación.

---

### Datos del proyecto

- **Estudiante:** Betiana Marquez
- **Proyecto:** Urban Import AI
- **Tecnología principal:** Python + Inteligencia Artificial

# 1. Introducción

## 1.1 Problema a abordar

En una empresa dedicada a la importación y comercialización de
productos, los clientes pueden realizar consultas relacionadas con
productos, precios, disponibilidad y características.

Responder estas consultas de manera rápida, clara y consistente
puede representar un desafío cuando se utilizan modelos de
inteligencia artificial sin instrucciones suficientemente
estructuradas.

Un prompt demasiado general puede producir respuestas extensas,
poco precisas o incluir información que no fue proporcionada.

Por este motivo, Urban Import AI propone analizar cómo las técnicas
de Fast Prompting pueden mejorar la interacción entre un usuario y
un modelo de inteligencia artificial.

## 1.2 Relevancia

La aplicación de técnicas de prompting permite establecer con mayor
precisión el rol del modelo, el contexto de la consulta, las reglas
que debe seguir y el formato esperado de la respuesta.

Esto puede contribuir a obtener respuestas más consistentes y
reducir consultas innecesarias a la API.

## 1.3 Propuesta de solución

La propuesta consiste en desarrollar y comparar dos estrategias:

1. Un prompt básico, con instrucciones generales.
2. Un prompt optimizado utilizando técnicas de Fast Prompting.

Posteriormente se analizarán las diferencias entre ambas estrategias
y se evaluará cuál resulta más adecuada para Urban Import AI.


## Viabilidad técnica

La propuesta es técnicamente viable porque utiliza herramientas
disponibles y accesibles para desarrollar una prueba de concepto,
como Python, Jupyter Notebook, Visual Studio Code y GitHub.

### Recursos disponibles

Para desarrollar la POC se utilizaron los siguientes recursos:

- Python para la implementación de la lógica.
- Jupyter Notebook para documentar y ejecutar las pruebas.
- Visual Studio Code como entorno de desarrollo.
- GitHub para almacenar y publicar el proyecto.
- Un modelo de inteligencia artificial para generación de texto.
- Una herramienta de generación de imágenes mediante IA.

### Estimación del tiempo de desarrollo

El desarrollo de la POC puede dividirse en las siguientes etapas:

| Actividad | Tiempo estimado |
|---|---:|
| Análisis del problema | 1 hora |
| Diseño de prompts | 2 horas |
| Desarrollo de la simulación | 2 horas |
| Pruebas y comparación | 1 hora |
| Generación de imagen | 1 hora |
| Documentación y README | 2 horas |
| Revisión final | 1 hora |
| **Total estimado** | **10 horas** |

Estos valores representan una estimación para una prueba de concepto
académica y pueden variar según la experiencia del desarrollador y las
herramientas utilizadas.

### Riesgos identificados

Durante el desarrollo pueden presentarse diferentes riesgos:

1. **Respuestas incorrectas o información inventada:** un modelo de IA
   puede generar información que no fue proporcionada.

2. **Consumo de recursos de API:** el uso de modelos mediante API puede
   generar costos dependiendo de la cantidad de consultas y tokens
   utilizados.

3. **Limitaciones de herramientas gratuitas:** algunas herramientas de
   generación de imágenes pueden tener límites de uso.

4. **Variación en la calidad de las respuestas:** pequeños cambios en
   las instrucciones pueden modificar el resultado generado.

5. **Errores durante la implementación:** una configuración incorrecta
   del entorno puede impedir la ejecución de la POC.

### Medidas de mitigación

Para reducir estos riesgos se aplicaron diferentes estrategias:

- Definir claramente el rol y objetivo del asistente.
- Establecer reglas y restricciones dentro del prompt.
- Indicar al modelo que no invente información.
- Utilizar datos controlados para las pruebas.
- Comparar diferentes estrategias de prompting.
- Medir el tamaño de las instrucciones utilizadas.
- Documentar las limitaciones de la POC.

# Reflexión sobre la Preentrega 1

En la Preentrega 1 se planteó el desarrollo de una propuesta de
inteligencia artificial orientada a responder consultas de clientes
de Urban Import AI. La propuesta inicial permitió identificar el
problema y plantear una solución basada en el uso de prompts.

A partir de esta propuesta inicial, en la presente entrega se
profundizó el diseño de los prompts mediante la aplicación de técnicas
de Fast Prompting.

La principal mejora consiste en pasar de una instrucción general a una
estructura organizada que define el rol del asistente, el objetivo,
el contexto, las reglas, las restricciones y el formato esperado de
la respuesta.

Esta estructura permite tener un mayor control sobre el comportamiento
esperado del modelo y reducir la posibilidad de que genere información
que no haya sido proporcionada.

Además, la propuesta final incorpora una comparación entre el prompt
básico y el prompt optimizado, un análisis de costos y consumo de
recursos, y la incorporación de un modelo de generación de imágenes.

De esta manera, la propuesta desarrollada en la Preentrega 1 se
mantiene como base del proyecto, pero se amplía mediante técnicas de
Fast Prompting y nuevas funcionalidades para lograr una prueba de
concepto más completa.


# 2. Objetivos

## Objetivo general

Desarrollar una prueba de concepto de Urban Import AI utilizando
técnicas de Fast Prompting para mejorar la calidad, claridad y
consistencia de las respuestas generadas por un asistente de
inteligencia artificial.

## Objetivos específicos

- Diseñar un prompt básico para responder consultas de clientes.
- Diseñar un prompt optimizado utilizando técnicas de prompting.
- Comparar los resultados obtenidos.
- Reducir respuestas ambiguas o innecesariamente extensas.
- Evitar que el modelo invente información no proporcionada.
- Crear una estructura de prompt reutilizable.
- Minimizar la cantidad de consultas realizadas a la API.
- Analizar la viabilidad de la propuesta.
 

# 3. Metodología

El proyecto se desarrollará mediante las siguientes etapas:

### Etapa 1 - Definición del problema

Se identifica la necesidad de mejorar la forma en que un asistente
virtual responde consultas relacionadas con productos de importación.

### Etapa 2 - Prompt básico

Se desarrolla una instrucción simple para observar el comportamiento
inicial del modelo.

### Etapa 3 - Prompt optimizado

Se incorporan técnicas de Fast Prompting:

- Definición del rol.
- Contexto.
- Objetivo.
- Instrucciones específicas.
- Restricciones.
- Formato de respuesta.
- Manejo de información desconocida.

### Etapa 4 - Comparación

Se utilizan consultas similares para analizar las diferencias entre
el prompt básico y el prompt optimizado.

### Etapa 5 - Evaluación

Los resultados se analizan considerando:

- Claridad.
- Relevancia.
- Consistencia.
- Control de información.
- Cantidad de consultas.
- Costo potencial.

### Etapa 6 - Conclusiones

Finalmente se determina si las técnicas de Fast Prompting permiten
mejorar la propuesta original.

# 4. Herramientas y tecnologías

## Python

Se utilizará Python como lenguaje principal para desarrollar la
prueba de concepto y realizar las pruebas con los prompts.

## Jupyter Notebook

Jupyter Notebook permitirá combinar explicaciones, código, resultados
y conclusiones dentro de un mismo documento interactivo.

## Modelo de inteligencia artificial

Se utilizará un modelo de lenguaje para evaluar el comportamiento de
los diferentes prompts.

La conexión con la API se realizará mediante la biblioteca oficial
correspondiente.

## Fast Prompting

Las principales técnicas utilizadas serán:

### Definición de rol

Se establecerá el rol que debe asumir el asistente.

### Contexto

Se proporcionará información relevante sobre Urban Import AI y el
tipo de consultas que debe responder.

### Instrucciones específicas

Se indicará claramente qué debe hacer el modelo.

### Restricciones

Se establecerán límites para evitar que el modelo invente
información.

### Formato de respuesta

Se definirá una estructura para obtener respuestas ordenadas y
fáciles de interpretar.

### Reutilización del prompt

Se diseñará una estructura que pueda utilizarse para diferentes
consultas sin tener que crear un prompt completamente nuevo cada vez.

## GitHub

GitHub será utilizado para almacenar y publicar el proyecto final,
incluyendo la notebook, documentación y archivos necesarios para
reproducir la POC.

# 5. Implementación

En esta etapa se desarrollará la prueba de concepto de Urban Import AI.

La implementación permitirá comparar un prompt básico con un prompt
optimizado mediante técnicas de Fast Prompting.

Para evitar consultas innecesarias y costos asociados al uso de una
API, inicialmente se trabajará con una simulación local.

La estructura desarrollada posteriormente podrá conectarse a un modelo
de inteligencia artificial mediante una API.

## 5.1 Prompt básico

El primer prompt utiliza una instrucción general para responder las
consultas del cliente.

### Prompt

> Sos un asistente de Urban Import AI. Respondé las consultas de los
> clientes sobre productos de importación de forma clara y amable.

## 5.2 Prompt optimizado mediante Fast Prompting

El segundo prompt incorpora una estructura más detallada.

Se define el rol del asistente, el contexto, el objetivo, las
restricciones y el formato esperado de respuesta.

### Prompt optimizado

> **ROL:**  
> Sos el asistente virtual de Urban Import AI, especializado en
> atención al cliente y productos de importación.
>
> **OBJETIVO:**  
> Responder las consultas de los clientes de manera clara, breve,
> precisa y cordial.
>
> **CONTEXTO:**  
> El cliente puede consultar sobre productos, características,
> disponibilidad y precios.
>
> **REGLAS:**  
> 1. Utilizá únicamente la información proporcionada.
> 2. No inventes precios, stock ni características.
> 3. Si falta información, indicá que no está disponible.
> 4. No agregues información que no haya sido solicitada.
> 5. Utilizá un lenguaje sencillo y profesional.
>
> **FORMATO:**  
> Presentá la respuesta de forma clara y organizada.
>
> **CONSULTA DEL CLIENTE:**  
> [consulta]
> 

In [1]:
# Prompt básico:
# contiene instrucciones generales para observar cómo responde
# el asistente cuando tiene poca información sobre el comportamiento esperado.
prompt_basico = """
Sos un asistente de Urban Import AI.
Respondé las consultas de los clientes sobre productos de importación
de forma clara y amable.
"""

# Prompt optimizado:
# incorpora técnicas de Fast Prompting como rol, objetivo,
# contexto, reglas, restricciones y formato de respuesta.
prompt_optimizado = """
ROL:
Sos el asistente virtual de Urban Import AI, especializado en
atención al cliente y productos de importación.

OBJETIVO:
Responder las consultas de los clientes de manera clara, breve,
precisa y cordial.

CONTEXTO:
El cliente puede consultar sobre productos, características,
disponibilidad y precios.

REGLAS:
1. Utilizá únicamente la información proporcionada.
2. No inventes precios, stock ni características.
3. Si falta información, indicá que no está disponible.
4. No agregues información que no haya sido solicitada.
5. Utilizá un lenguaje sencillo y profesional.

FORMATO:
Presentá la respuesta de forma clara y organizada.
"""

# Mostramos ambos prompts para poder compararlos visualmente
# antes de realizar la simulación.
print("PROMPT BÁSICO")
print("=" * 50)
print(prompt_basico)

print("\nPROMPT OPTIMIZADO")
print("=" * 50)
print(prompt_optimizado)

PROMPT BÁSICO

Sos un asistente de Urban Import AI.
Respondé las consultas de los clientes sobre productos de importación
de forma clara y amable.


PROMPT OPTIMIZADO

ROL:
Sos el asistente virtual de Urban Import AI, especializado en
atención al cliente y productos de importación.

OBJETIVO:
Responder las consultas de los clientes de manera clara, breve,
precisa y cordial.

CONTEXTO:
El cliente puede consultar sobre productos, características,
disponibilidad y precios.

REGLAS:
1. Utilizá únicamente la información proporcionada.
2. No inventes precios, stock ni características.
3. Si falta información, indicá que no está disponible.
4. No agregues información que no haya sido solicitada.
5. Utilizá un lenguaje sencillo y profesional.

FORMATO:
Presentá la respuesta de forma clara y organizada.



In [2]:
# Datos de prueba del producto.
# Estos datos simulan la información que tendría disponible
# un asistente de atención al cliente.
producto = {
    "nombre": "Auriculares Bluetooth Urban X1",
    "precio": "$45.000",
    "stock": "Disponible",
    "caracteristicas": [
        "Bluetooth 5.3",
        "Batería de hasta 20 horas",
        "Micrófono integrado"
    ]
}

# Consulta que realizará el cliente al asistente.
consulta = "¿Cuánto cuestan los auriculares Urban X1 y tienen stock?"

## 5.3 Simulación de resultados

Para demostrar el funcionamiento de la POC sin realizar llamadas
innecesarias a una API, se utiliza un conjunto de datos de prueba.

La información disponible sobre el producto es:

- **Producto:** Auriculares Bluetooth Urban X1
- **Precio:** $45.000
- **Stock:** Disponible
- **Bluetooth:** 5.3
- **Batería:** Hasta 20 horas
- **Micrófono:** Integrado

La consulta utilizada será:

> ¿Cuánto cuestan los auriculares Urban X1 y tienen stock?
> 

In [3]:
# Construimos una respuesta utilizando los datos disponibles
# del producto. Al utilizar la información del diccionario,
# evitamos escribir manualmente los datos en diferentes lugares.
respuesta_optimizada = f"""
Los Auriculares Bluetooth Urban X1 tienen un precio de {producto["precio"]}.

Stock: {producto["stock"]}.

Características principales:
- Bluetooth 5.3
- Hasta 20 horas de batería
- Micrófono integrado
"""

# Mostramos el resultado final de la simulación.
print(respuesta_optimizada)


Los Auriculares Bluetooth Urban X1 tienen un precio de $45.000.

Stock: Disponible.

Características principales:
- Bluetooth 5.3
- Hasta 20 horas de batería
- Micrófono integrado



# 6. Comparación de estrategias

La comparación permite observar la diferencia entre utilizar una
instrucción general y una estructura de prompt más específica.

| Criterio | Prompt básico | Prompt optimizado |
|---|---|---|
| Definición de rol | Básica | Específica |
| Contexto | Limitado | Definido |
| Restricciones | No | Sí |
| Formato de respuesta | No definido | Definido |
| Control de información | Bajo | Mayor |
| Reutilización | Media | Alta |
| Claridad esperada | Media | Alta |

El prompt optimizado permite controlar mejor el comportamiento esperado
del asistente porque establece explícitamente el rol, el objetivo,
el contexto y las restricciones.

Además, evita que el modelo agregue información que no haya sido
proporcionada.

# 7. Evaluación de la POC

La prueba de concepto permite observar que un prompt estructurado
ofrece un mayor control sobre las respuestas esperadas.

Las principales mejoras obtenidas mediante Fast Prompting son:

- Mayor claridad en las instrucciones.
- Definición explícita del rol del asistente.
- Incorporación de contexto.
- Establecimiento de reglas y restricciones.
- Reducción de respuestas innecesarias.
- Mayor consistencia entre diferentes consultas.

# 8. Análisis de costos y optimización

Uno de los objetivos del proyecto es minimizar las consultas realizadas
a la API.

La POC fue diseñada para evitar llamadas innecesarias al modelo.

En una implementación real se recomienda:

- Utilizar una única consulta cuando sea suficiente.
- Evitar repetir información dentro del prompt.
- Mantener las instrucciones claras y concisas.
- No realizar varias consultas para resolver una misma pregunta.
- Utilizar respuestas estructuradas para facilitar el procesamiento.

Durante esta demostración se utiliza una simulación local, por lo que
no se generan costos de API.

## 8.1 Medición del tamaño de los prompts

Para analizar de forma más concreta el impacto de las instrucciones
utilizadas, se comparará la cantidad de palabras y caracteres presentes
en el prompt básico y en el prompt optimizado.

Esta medición permite realizar una aproximación al volumen de
información que debería procesar un modelo de inteligencia artificial.

Es importante aclarar que la cantidad de caracteres o palabras no
equivale exactamente a la cantidad de tokens. Sin embargo, permite
realizar una primera comparación cuantitativa entre ambas estrategias.

In [4]:
# Medimos la cantidad de caracteres y palabras de cada prompt.
# Esta comparación permite observar la diferencia de tamaño
# entre el prompt básico y el prompt optimizado.

caracteres_basico = len(prompt_basico)
palabras_basico = len(prompt_basico.split())

caracteres_optimizado = len(prompt_optimizado)
palabras_optimizado = len(prompt_optimizado.split())

print("PROMPT BÁSICO")
print("Caracteres:", caracteres_basico)
print("Palabras:", palabras_basico)

print("\nPROMPT OPTIMIZADO")
print("Caracteres:", caracteres_optimizado)
print("Palabras:", palabras_optimizado)

PROMPT BÁSICO
Caracteres: 133
Palabras: 22

PROMPT OPTIMIZADO
Caracteres: 639
Palabras: 91


## 8.2 Interpretación de la medición

La medición permite observar que el prompt optimizado contiene más
información que el prompt básico debido a que incorpora contexto,
objetivos, reglas y restricciones.

Esto demuestra que optimizar un prompt no significa necesariamente
hacerlo más corto.

El objetivo de Fast Prompting es utilizar instrucciones claras y
estructuradas para obtener mejores resultados y evitar consultas
innecesarias o respuestas que requieran nuevas interacciones.

Por este motivo, el análisis de eficiencia debe considerar tanto el
tamaño de las instrucciones como la cantidad de consultas realizadas y
la calidad de las respuestas obtenidas.



# 9. Generación de imágenes mediante IA

Como complemento al modelo de generación de texto, la propuesta
incorpora un modelo de inteligencia artificial capaz de generar
imágenes a partir de instrucciones escritas.

Esta funcionalidad permite ampliar la aplicación de Urban Import AI
hacia la creación de contenido visual para productos.

El proceso utilizado consiste en definir las características del
producto, construir un prompt descriptivo y utilizar un generador de
imágenes para obtener una representación visual.



# 9. Análisis de costos y optimización

El uso de modelos de inteligencia artificial puede generar costos
relacionados con la cantidad de información procesada y la cantidad
de consultas realizadas.

Por este motivo, en esta POC se analiza el tamaño de los prompts y se
realiza una medición local de palabras, caracteres y tokens.

## 9.1 Medición del tamaño de los prompts

Para analizar de forma más concreta el impacto de las instrucciones
utilizadas, se comparará la cantidad de palabras y caracteres presentes
en el prompt básico y en el prompt optimizado.

Esta medición permite realizar una aproximación al volumen de
información que debería procesar un modelo de inteligencia artificial.

Es importante aclarar que la cantidad de caracteres o palabras no
equivale exactamente a la cantidad de tokens. Por este motivo, también
se realizará una medición local de tokens.



## 9.3 Medición de tokens

Además de comparar caracteres y palabras, se realiza una estimación
del número de tokens utilizados por cada prompt.

La cantidad de tokens es una medida más relacionada con el procesamiento
de texto de los modelos de inteligencia artificial y permite realizar
un análisis más preciso del posible consumo de recursos.

Para esta prueba se utiliza la librería `tiktoken`, que permite
realizar una tokenización local sin necesidad de realizar llamadas
a una API.

In [5]:
import tiktoken

# Utilizamos una codificación compatible con modelos de OpenAI.
encoding = tiktoken.get_encoding("cl100k_base")

# Contamos los tokens de cada prompt.
tokens_basico = len(encoding.encode(prompt_basico))
tokens_optimizado = len(encoding.encode(prompt_optimizado))

print("PROMPT BÁSICO")
print("Tokens:", tokens_basico)

print("\nPROMPT OPTIMIZADO")
print("Tokens:", tokens_optimizado)

# Calculamos la diferencia entre ambos prompts.
diferencia_tokens = tokens_optimizado - tokens_basico

print("\nDiferencia de tokens:", diferencia_tokens)

PROMPT BÁSICO
Tokens: 34

PROMPT OPTIMIZADO
Tokens: 156

Diferencia de tokens: 122


In [8]:
%pip install tiktoken

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


### Interpretación de los resultados de tokens

La medición realizada muestra que el prompt básico utiliza 34 tokens,
mientras que el prompt optimizado utiliza 156 tokens.

Esto representa una diferencia de 122 tokens entre ambas estrategias.

El prompt optimizado utiliza más tokens porque incorpora información
adicional relacionada con el rol del asistente, el objetivo, el
contexto, las reglas, las restricciones y el formato esperado de la
respuesta.

Este resultado permite observar que la optimización mediante Fast
Prompting no significa necesariamente utilizar menos tokens.

En este caso, se utiliza una mayor cantidad de tokens para proporcionar
instrucciones más completas y controlar mejor el comportamiento
esperado del asistente.

En una implementación real mediante una API, una mayor cantidad de
tokens podría representar un mayor costo de procesamiento. Sin
embargo, el análisis de eficiencia debe considerar también la calidad
de las respuestas y la cantidad de consultas necesarias para resolver
una solicitud.

Por lo tanto, los resultados obtenidos muestran la importancia de
buscar un equilibrio entre la cantidad de instrucciones proporcionadas,
la calidad de las respuestas y el consumo de recursos.

## 9.4 Simulación de cantidad de consultas

Además del consumo de tokens, es importante considerar la cantidad de
consultas necesarias para resolver una solicitud.

Para demostrar este aspecto, se realiza una simulación sencilla
comparando una estrategia básica con una estrategia optimizada.

La estrategia básica puede requerir consultas adicionales cuando la
información necesaria no está contemplada inicialmente.

La estrategia optimizada incorpora desde el comienzo el contexto,
las reglas y las restricciones necesarias para responder la consulta.

In [7]:
# Simulación de cantidad de consultas necesarias.
# Los valores representan un escenario de prueba de la POC.

consultas_basico = 3
consultas_optimizado = 1

reduccion_consultas = consultas_basico - consultas_optimizado

print("ESTRATEGIA BÁSICA")
print("Consultas realizadas:", consultas_basico)

print("\nESTRATEGIA OPTIMIZADA")
print("Consultas realizadas:", consultas_optimizado)

print("\nReducción de consultas:", reduccion_consultas)

ESTRATEGIA BÁSICA
Consultas realizadas: 3

ESTRATEGIA OPTIMIZADA
Consultas realizadas: 1

Reducción de consultas: 2


## 9.5 Conclusión del análisis de costos

El análisis realizado permite comparar diferentes aspectos relacionados
con la eficiencia de los prompts.

En la medición realizada, el prompt básico utilizó 34 tokens, mientras
que el prompt optimizado utilizó 156 tokens. La diferencia fue de 122
tokens.

Esto demuestra que el prompt optimizado requiere más información para
establecer el rol, contexto, reglas, restricciones y formato de
respuesta.

Por otro lado, la simulación de consultas mostró una reducción de 3
consultas en la estrategia básica a 1 consulta en la estrategia
optimizada.

Los valores de consultas corresponden a una simulación y no a llamadas
reales a una API.

En consecuencia, la eficiencia de una estrategia de Fast Prompting no
debe analizarse únicamente por la cantidad de tokens utilizados, sino
también por la cantidad de interacciones necesarias y por la calidad y
consistencia de las respuestas obtenidas.

Una implementación futura mediante una API permitiría medir estos
factores con datos reales de consumo, tiempo de respuesta y costos.

## 10.1 Prompt utilizado para la generación de imagen

Para demostrar el funcionamiento del modelo texto → imagen se utilizará
como ejemplo el producto Auriculares Bluetooth Urban X1.

El prompt fue diseñado utilizando una estructura descriptiva que
especifica el producto, el contexto visual, el estilo y el objetivo de
la imagen.

### Prompt

> Crear una imagen publicitaria profesional de unos auriculares
> Bluetooth modernos llamados Urban X1, de diseño elegante y
> minimalista, colocados sobre una superficie limpia.
>
> La imagen debe tener iluminación de estudio, fondo moderno y
> atractivo, composición comercial y aspecto de fotografía profesional
> de producto.
>
> La imagen debe transmitir tecnología, calidad y modernidad, y estar
> pensada para utilizarse en una tienda online.

## 10.2 Resultado de la generación

El prompt anterior se utilizó en una herramienta de generación de
imágenes mediante inteligencia artificial.

El resultado obtenido se incorporará a continuación como evidencia de
la utilización de un modelo texto → imagen.

### Imagen generada

La siguiente imagen fue generada a partir del prompt anterior
utilizando un modelo de inteligencia artificial de generación
de imágenes.

![Auriculares Urban X1](images/urban_x1_generada.png)

La generación de esta imagen demuestra la utilización de un modelo
texto → imagen como complemento del modelo texto → texto utilizado
en la POC.

## 10.3 Evaluación del resultado

El resultado obtenido presenta el producto Urban X1 mediante una
composición visual orientada al comercio electrónico.

La imagen incorpora características solicitadas en el prompt, como
un diseño moderno, iluminación de estudio, presentación profesional
del producto y elementos visuales relacionados con tecnología.

La generación texto → imagen complementa la funcionalidad texto →
texto de Urban Import AI, ya que permite utilizar la inteligencia
artificial no solamente para responder consultas, sino también para
crear contenido visual relacionado con los productos.

### Criterios de evaluación

| Criterio | Resultado observado |
|---|---|
| Representación del producto | El producto aparece claramente identificado |
| Calidad visual | Presentación con aspecto comercial |
| Cumplimiento del prompt | Se incorporan las características visuales solicitadas |
| Utilidad comercial | Puede utilizarse como material visual para una tienda online |
| Complemento del modelo texto → texto | Permite ampliar la POC hacia contenido visual |

# 11. Generación de imágenes mediante IA

Como complemento al modelo de generación de texto, la propuesta
incorpora un modelo de inteligencia artificial capaz de generar
imágenes a partir de instrucciones escritas.

Esta funcionalidad permite ampliar la aplicación de Urban Import AI
hacia la creación de contenido visual para productos.

El proceso utilizado consiste en definir las características del
producto, construir un prompt descriptivo y utilizar un generador de
imágenes para obtener una representación visual.

## 11.1 Prompt utilizado para la generación de imagen

Para demostrar el funcionamiento del modelo texto → imagen se utilizará
como ejemplo el producto Auriculares Bluetooth Urban X1.

El prompt fue diseñado utilizando una estructura descriptiva que
especifica el producto, el contexto visual, el estilo y el objetivo de
la imagen.

### Prompt

> Crear una imagen publicitaria profesional de unos auriculares
> Bluetooth modernos llamados Urban X1, de diseño elegante y
> minimalista, colocados sobre una superficie limpia.
>
> La imagen debe tener iluminación de estudio, fondo moderno y
> atractivo, composición comercial y aspecto de fotografía profesional
> de producto.
>
> La imagen debe transmitir tecnología, calidad y modernidad, y estar
> pensada para utilizarse en una tienda online.

## 11.2 Resultado de la generación

El prompt anterior se utilizó en una herramienta de generación de
imágenes mediante inteligencia artificial.

El resultado obtenido se incorporará a continuación como evidencia de
la utilización de un modelo texto → imagen.

### Imagen generada

La siguiente imagen fue generada a partir del prompt anterior
utilizando un modelo de inteligencia artificial de generación
de imágenes.

![Auriculares Urban X1](images/urban_x1_generada.png)

La generación de esta imagen demuestra la utilización de un modelo
texto → imagen como complemento del modelo texto → texto utilizado
en la POC.

## 11.3 Evaluación del resultado

El resultado obtenido presenta el producto Urban X1 mediante una
composición visual orientada al comercio electrónico.

La imagen incorpora características solicitadas en el prompt, como
un diseño moderno, iluminación de estudio, presentación profesional
del producto y elementos visuales relacionados con tecnología.

La generación texto → imagen complementa la funcionalidad texto →
texto de Urban Import AI, ya que permite utilizar la inteligencia
artificial no solamente para responder consultas, sino también para
crear contenido visual relacionado con los productos.

### Criterios de evaluación

| Criterio | Resultado observado |
|---|---|
| Representación del producto | El producto aparece claramente identificado |
| Calidad visual | Presentación con aspecto comercial |
| Cumplimiento del prompt | Se incorporan las características visuales solicitadas |
| Utilidad comercial | Puede utilizarse como material visual para una tienda online |
| Complemento del modelo texto → texto | Permite ampliar la POC hacia contenido visual |

# 12. Conclusiones

El desarrollo de esta prueba de concepto permitió aplicar técnicas de
Fast Prompting a un caso de uso concreto: la atención de consultas de
clientes de Urban Import AI.

La comparación entre un prompt básico y un prompt optimizado permitió
observar la importancia de estructurar correctamente las instrucciones
proporcionadas al modelo.

La utilización de un rol definido, contexto, objetivos, reglas,
restricciones y formato de respuesta permite establecer un mayor
control sobre el comportamiento esperado del asistente.

Además, la medición del tamaño de los prompts permitió realizar una
comparación cuantitativa entre ambas estrategias. Esta medición también
permitió observar que optimizar un prompt no significa necesariamente
hacerlo más corto, sino utilizar instrucciones claras y estructuradas
para obtener mejores resultados y evitar interacciones innecesarias.

La incorporación de un modelo texto → imagen permitió ampliar la
propuesta original, demostrando que la inteligencia artificial puede
utilizarse tanto para generar respuestas de texto como para crear
contenido visual relacionado con los productos.

La POC desarrollada constituye una base para futuras mejoras, como la
conexión con una API, la incorporación de una base de datos de
productos, la medición de tokens y costos reales y la implementación de
una interfaz interactiva.

# 13 Trabajo futuro

Como posibles mejoras y ampliaciones del proyecto se plantean las
siguientes:

1. Conectar la notebook con un modelo de inteligencia artificial
   mediante una API para realizar pruebas con respuestas generadas
   directamente por el modelo.

2. Incorporar una interfaz interactiva que permita a los usuarios
   ingresar consultas sobre los productos.

3. Integrar una base de datos de productos con información actualizada
   sobre precios, stock y características.

4. Medir de forma directa la cantidad de tokens utilizados, el tiempo
   de respuesta y el costo de las consultas realizadas a una API.

5. Comparar diferentes modelos de inteligencia artificial y distintas
   estrategias de prompting.

6. Incorporar un sistema de evaluación automática para analizar la
   calidad de las respuestas generadas.

7. Ampliar la generación de imágenes para crear contenido visual de
   diferentes productos.

8. Integrar las funcionalidades de generación de texto e imágenes en
   una solución más completa para la atención y comercialización de
   productos de Urban Import AI.

# 14. Referencias

- Material de estudio y contenidos proporcionados durante el curso
  relacionados con Inteligencia Artificial y Fast Prompting.

- Documentación oficial de Jupyter Notebook para la ejecución y
  documentación de código Python.

- Documentación de GitHub para el almacenamiento y publicación de
  proyectos.

- Documentación de las herramientas de inteligencia artificial
  utilizadas para la generación de texto e imágenes.

- Recursos educativos consultados sobre diseño y optimización de
  prompts.